extensions-contrib/materialized-view-selection/src/main/java/io/druid/query/materializedview/DerivativesManager.java

```diff
@@ -0,0 +1,246 @@
+import com.google.common.util.concurrent.ListeningScheduledExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.inject.Inject;
+import io.druid.guice.ManageLifecycleLast;
+import io.druid.indexing.overlord.DataSourceMetadata;
+import io.druid.metadata.MetadataStorageTablesConfig;
+import io.druid.metadata.SQLMetadataConnector;
+import io.druid.indexing.materializedview.DerivativeDataSourceMetadata;
+import io.druid.java.util.common.DateTimes;
+import io.druid.java.util.common.Intervals;
+import io.druid.java.util.common.Pair;
+import io.druid.java.util.common.StringUtils;
+import io.druid.java.util.common.concurrent.Execs;
+import io.druid.java.util.common.lifecycle.LifecycleStart;
+import io.druid.java.util.common.lifecycle.LifecycleStop;
+import io.druid.java.util.emitter.EmittingLogger;
+import io.druid.timeline.DataSegment;
+import org.joda.time.Duration;
+import org.joda.time.Interval;
+import org.skife.jdbi.v2.Handle;
+import org.skife.jdbi.v2.StatementContext;
+import org.skife.jdbi.v2.tweak.HandleCallback;
+import org.skife.jdbi.v2.tweak.ResultSetMapper;
+
+import java.io.IOException;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Read and store derivatives information from dataSource table frequently.
+ * When optimize query, DerivativesManager offers the information about derivatives.
+ */
+@ManageLifecycleLast
+public class DerivativesManager 
+{
+  private static final EmittingLogger log = new EmittingLogger(DerivativesManager.class);
+  private final MaterializedViewConfig config;
+  private final Supplier<MetadataStorageTablesConfig> dbTables;
+  private final SQLMetadataConnector connector;
+  private final ObjectMapper objectMapper;
+  private final Object lock = new Object();
+  
+  private static volatile AtomicReference<ConcurrentHashMap<String, SortedSet<Derivative>>> derivativesRef = new AtomicReference<>(new ConcurrentHashMap<>());
+  private volatile boolean started = false;
+  private volatile ListeningScheduledExecutorService exec = null;
+  private volatile ListenableFuture<?> future = null;
+  
+  @Inject
+  public DerivativesManager(
+      MaterializedViewConfig config,
+      Supplier<MetadataStorageTablesConfig> dbTables,
+      ObjectMapper objectMapper,
+      SQLMetadataConnector connector)
+  {
+    this.config = config;
+    this.dbTables = dbTables;
+    this.objectMapper = objectMapper;
+    this.connector = connector;
+  }
+
+  @LifecycleStart
+  public void start()
+  {
+    log.info("starting derivatives manager.");
+    synchronized (lock) {
+      if (started) {
+        return;
+      }
+      exec = MoreExecutors.listeningDecorator(Execs.scheduledSingleThreaded("SQLMetadataDerivativesManager-Exec--%d"));
```

Comment:
The thread name is usually useful when debugging what a thread is for. Probably `DerivativesManager-exec-%d` is a better name.